### ISSUE TITLE
Color contrast warning appearing incorrectly

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#43537]
### Description

After setting color options for a nested block, the parent block seems to start showing the color contrast warning, even if none exist. I am still working out the circumstances for when this happens but have narrowed it down to at least one flow. 

### Step-by-step reproduction instructions

1. Add a query loop block.
2. Change the colors of the post title block within it.
3. Notice that the color contrast warning appears when selecting the overall query loop block.

### Screenshots, screen recording, code snippet

Here's me talking through the problem:

https://user-images.githubusercontent.com/26996883/186230569-67ee6fec-e49c-4561-bf32-5b5378daa3f9.mov

### Environment info

- WP 6.0.1
- Gutenberg 13.9
- TT2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by glendaviesnz** [comment]
In the example shown above this is due to the link text being set to white on the post title block, and the query block having a default background color of white. 

To check the contrast of link text against block background the contrast [checker looks for the first child `a`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/color-panel.js#L40). 

When you navigate back up to the parent query block, even though no color is set on the block it gets the computed color value from the block element, which in the above case is white, and then looks for the first `a`, which happens to be the post title link which was explicitly set to white, hence the warning as it ignores the fact that the post title block has a green background set. 

The funny thing is if you set the Post title link color to green, which makes it disappear, the contrast warning at the query block level disappears as it now thinks you have dark green text on a white background 😄 

This is a tricky one as it is not really possible to know which child block backgrounds may affect the link contrast - I wonder if the link color check should be narrowed down to direct children of the parent block?

**Comment by jasmussen** [comment]
Thanks for the thorough context Glen!

> I wonder if the link color check should be narrowed down to direct children of the parent block?

It does seem mainly a UI issue. Essentially the color warning should only ever be shown in context of the problem source. I.e. when all of the colors are unset, you are not looking at the source of the problem and while you could take action there, it would only fix the problem locally. 

> even though no color is set on the block it gets the computed color value from the block element, which in the above case is white, and then looks for the first `a`

This makes it sound like the contrast warning should appear in Global Styles rather than locally here? Let me know if I missed anything here, it's a gnarly one!

**Comment by glendaviesnz** [comment]
One [possible fix for this specific problem here ](https://github.com/WordPress/gutenberg/pull/43592) ... but it is at the expense of missing contrast issues when colors are not explicitly set at the parent level and a link nested more than one level deep does have bad contrast. However, explicitly setting bad contrast colors on the block should still work, so it may be less confusing to not show a warning for non explicit contrast issues on deeply nested children, than to show the false positives as in @annezazu example.

**Comment by jasmussen** [comment]
Nice fix. As I noted on the PR, it seems most correct to only show the contrast warning in context of explicitly set colors, as that would also be the best place to fix any color issue. Otherwise you might end up trying to fix a contrast issue in the wrong place.
